File: tests/meta_order_test.php
<?php
/**
 * The panel half of the metadata fetch order: includes/meta_order.php.
 *
 * The worker's half is covered by tests/meta_order_test.py, which also checks the two halves agree
 * on the list of selectors. This suite is about the one thing the panel is responsible for: what
 * comes out of the form always adds up to exactly 100, whatever went in. The worker re-normalises
 * too, but a queue running on shares that total 97 would be a silent 3 % of nothing.
 */
if (PHP_SAPI !== 'cli') { http_response_code(404); exit; }
require_once dirname(__DIR__) . '/includes/meta_order.php';

$fails = 0; $n = 0;
function check(string $name, bool $ok, string $info = ''): void {
    global $fails, $n;
    $n++;
    echo ($ok ? 'PASS ' : 'FAIL ') . $name . ($ok || $info === '' ? '' : '  -> ' . $info) . "\n";
    if (!$ok) $fails++;
}

// ── the mode ─────────────────────────────────────────────────────────────────
check('a known mode survives', metaOrderNormalise('seeders', [])[0] === 'seeders');
check('case does not matter', metaOrderNormalise('SEEDERS', [])[0] === 'seeders');
check('surrounding space does not matter', metaOrderNormalise('  mix ', [])[0] === 'mix');
check('an unknown mode becomes the default', metaOrderNormalise('sideways', [])[0] === 'oldest');
check('an empty mode becomes the default', metaOrderNormalise('', [])[0] === 'oldest');
foreach (metaOrderModes() as $m) {
    check('mode "' . $m . '" is accepted', metaOrderNormalise($m, [])[0] === $m);
}

// ── the whitelist share, which means something different at zero ─────────────
// 0 is not "never": it is "keep absolute priority", the behaviour of every release before this one.
// A test for it exists because that is the kind of meaning a later refactor flattens into "off".
[, $wl0] = metaOrderNormalise('mix', ['whitelist' => 0, 'seeders' => 100]);
check('a whitelist share of zero is preserved, not rounded away', $wl0['whitelist'] === 0, json_encode($wl0));
[, $wl30] = metaOrderNormalise('mix', ['whitelist' => 30, 'seeders' => 70]);
check('a whitelist share survives normalisation', $wl30['whitelist'] === 30, json_encode($wl30));
check('the shipped default leaves the whitelist on absolute priority',
      metaOrderDefaultMix()['whitelist'] === 0);

// ── the shares always total 100 ──────────────────────────────────────────────
// The property, checked over every shape a form can produce rather than over a handful of examples.
$cases = [
    'already correct'      => ['oldest' => 0, 'newest' => 15, 'seeders' => 70, 'random' => 15],
    'short of 100'         => ['oldest' => 10, 'newest' => 10, 'seeders' => 10, 'random' => 10],
    'over 100'             => ['oldest' => 90, 'newest' => 90, 'seeders' => 90, 'random' => 90],
    'one field only'       => ['seeders' => 40],
    'one field at 100'     => ['seeders' => 100],
    'awkward thirds'       => ['oldest' => 1, 'newest' => 1, 'seeders' => 1],
    'huge and tiny'        => ['seeders' => 999, 'random' => 1],
    'strings from a form'  => ['oldest' => '0', 'newest' => '15', 'seeders' => '70', 'random' => '15'],
    'a stray negative'     => ['seeders' => -20, 'newest' => 50],
    'nonsense beside real' => ['seeders' => 'lots', 'newest' => 30],
    'nothing at all'       => [],
];
foreach ($cases as $what => $in) {
    [, $out] = metaOrderNormalise('mix', $in);
    check('shares total exactly 100 (' . $what . ')', array_sum($out) === 100, json_encode($out));
    check('no share is out of range (' . $what . ')',
          count(array_filter($out, fn($v) => $v < 0 || $v > 100)) === 0, json_encode($out));
    check('every mix key is present (' . $what . ')',
          array_keys($out) === metaOrderMixKeys(), json_encode(array_keys($out)));
}

// Empty is not a configuration — it would leave the worker with nothing to rotate through, so it
// falls back to the shipped mix rather than silently becoming "longest waiting for ever".
check('every share at zero falls back to the shipped mix',
      metaOrderNormalise('mix', ['seeders' => 0])[1] === metaOrderDefaultMix());
check('the shipped mix itself totals 100', array_sum(metaOrderDefaultMix()) === 100);

// Normalising twice must not drift: the panel saves, reloads and saves again all the time.
[, $once] = metaOrderNormalise('mix', ['oldest' => 10, 'newest' => 10, 'seeders' => 10, 'random' => 10]);
[, $twice] = metaOrderNormalise('mix', $once);
check('normalising an already-normalised mix changes nothing', $once === $twice, json_encode([$once, $twice]));

// Proportions are kept, not flattened: 3:1 in must still be about 3:1 out.
[, $prop] = metaOrderNormalise('mix', ['seeders' => 30, 'newest' => 10]);
check('proportions survive rescaling', $prop['seeders'] === 75 && $prop['newest'] === 25, json_encode($prop));

// The rounding remainder goes to the largest share, where a point either way is least visible —
// and, more to the point, it never goes to a share the operator set to zero.
[, $thirds] = metaOrderNormalise('mix', ['oldest' => 1, 'newest' => 1, 'seeders' => 1]);
check('a share left at zero stays at zero after rescaling', $thirds['random'] === 0, json_encode($thirds));
check('the remainder lands on the largest share', max($thirds) === 34, json_encode($thirds));

// ── the definitions the rest of the panel builds on ──────────────────────────
check('there are six selectors', count(metaOrderSelectors()) === 6, json_encode(metaOrderSelectors()));
check('the mix distributes over the selectors plus the whitelist',
      metaOrderMixKeys() === array_merge(['whitelist'], metaOrderSelectors()), json_encode(metaOrderMixKeys()));
// `whitelist` is a share and NOT a mode: outside the mix the whitelist always drains first, and
// there is nothing in the index to sort by it — a whitelisted hash is deleted from the index on the
// next poll. Offering it as a mode would be offering an ordering of an empty set.
check('whitelist is not offered as a mode', !in_array('whitelist', metaOrderModes(), true));
// A lookup table, so the KEYS must match the mix keys but the order of a map is not meaningful —
// comparing sequences here would fail on a reordering that changes nothing.
check('every ordering names the index it rides on',
      count(array_diff(metaOrderMixKeys(), array_keys(metaOrderIndexes()))) === 0
      && count(array_diff(array_keys(metaOrderIndexes()), metaOrderMixKeys())) === 0,
      json_encode(array_keys(metaOrderIndexes())));
check('only random and whitelist need no secondary index',
      array_keys(array_filter(metaOrderIndexes(), fn($v) => $v === null)) === ['random', 'whitelist'],
      json_encode(metaOrderIndexes()));
// An index named here that the schema never creates would leave the worker refusing that selector
// for ever, and the panel showing "building" that never finishes.
$schemaSrc = (string)file_get_contents(dirname(__DIR__) . '/includes/schema.php');
foreach (metaOrderIndexes() as $sel => $idx) {
    if ($idx === null) continue;
    check("schema.php creates $idx (needed by \"$sel\")", str_contains($schemaSrc, '`' . $idx . '`'));
}
check('the orderings left out are written down with a reason', count(metaOrderRejected()) >= 3);
foreach (metaOrderRejected() as $rk => $rv) {
    check("the reason given for leaving out \"$rk\" is a sentence, not a shrug", strlen($rv) > 40);
}
check('the modes are the selectors plus "mix"',
      metaOrderModes() === array_merge(metaOrderSelectors(), ['mix']), json_encode(metaOrderModes()));
check('every mode has a label',
      array_keys(metaOrderModeLabels()) === metaOrderModes(), json_encode(array_keys(metaOrderModeLabels())));
check('every mix key has a share label',
      count(array_diff(metaOrderMixKeys(), array_keys(metaOrderShareLabels()))) === 0);
check('the default mix names exactly the mix keys',
      array_keys(metaOrderDefaultMix()) === metaOrderMixKeys(), json_encode(array_keys(metaOrderDefaultMix())));
check('the rotation is 100 claims, so one point is one claim', META_ORDER_ROTATION === 100);

// ── every key is registered where it has to be ───────────────────────────────
// Four places, or the setting saves cleanly and does nothing. This project has shipped that bug.
$root = dirname(__DIR__);
$schema = (string)file_get_contents($root . '/includes/schema.php');
$save   = (string)file_get_contents($root . '/api/admin/save_settings.php');
$cat    = (string)file_get_contents($root . '/includes/settings_catalog.php');
$tpl    = (string)file_get_contents($root . '/templates/admin/settings.php');
$keys = ['meta_order_mode'];
foreach (metaOrderMixKeys() as $sel) $keys[] = 'meta_order_mix_' . $sel;
foreach ($keys as $k) {
    check($k . ': has a schema default', str_contains($schema, "'" . $k . "'"));
    check($k . ': is in the save allow-list', str_contains($save, "'" . $k . "'"));
    check($k . ': is searchable', str_contains($cat, "'" . $k . "'"));
}
check('the form is built from this file, not a second copy of the list',
      str_contains($tpl, 'metaOrderModeLabels()') && str_contains($tpl, 'metaOrderShareLabels()'));
check('the save path normalises through this file', str_contains($save, 'metaOrderNormalise('));
// A settings-only change still needs a schema bump: the default rows are written by the migration
// block, which only runs when the VERSION moves.
check('the schema version was bumped for these defaults',
      preg_match('/TRACKER_SCHEMA_VERSION = (\d+)/', $schema, $m) && (int)$m[1] >= 31, $m[1] ?? '?');

echo "\n$n checks, $fails failed\n";
exit($fails ? 1 : 0);

File: tests/meta_order_test.py
#!/usr/bin/env python3
"""The metadata worker's fetch order — the rotation, the queue split, and the SQL each mode runs.

Why this suite exists at all: the index queue is ~3 million rows and resolves at a few per second,
so the ORDER of that queue decides what the tracker knows anything about for months. Three things
can go wrong quietly.

  * A rotation that is arithmetically right but BLOCKED (seventy of one kind, then fifteen of the
    next) makes each wave of parallel fetches a single kind and the balance only appears over hours —
    the numbers in the panel would still read 70/15/15.
  * A selector whose ORDER BY no index can serve turns every claim into a filesort over three
    million rows, several times a second. That looks like a slow machine, not a wrong setting.
  * The whitelist — rows a person asked for by name — silently losing its absolute priority.

All three are tested here by property, not by eye.

    python tests/meta_order_test.py
"""
import os, re, sys, types

sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)), '..', 'worker'))

# worker.py imports libtorrent and pymysql at module level and exits if either is missing. Neither
# has anything to do with the order of a queue, so they are stubbed: the alternative is a suite that
# only runs on the production machine, which is the one place it is least useful.
for name, attrs in (('libtorrent', {'__version__': '0.0-stub'}), ('pymysql', {})):
    if name not in sys.modules:
        mod = types.ModuleType(name)
        for k, v in attrs.items():
            setattr(mod, k, v)
        if name == 'pymysql':
            mod.cursors = types.SimpleNamespace(DictCursor=object)
            mod.err = types.SimpleNamespace(OperationalError=Exception, InterfaceError=Exception)
        sys.modules[name] = mod

import worker as W  # noqa: E402

fails = 0
n = 0


def check(name, ok, info=''):
    global fails, n
    n += 1
    print(('PASS ' if ok else 'FAIL ') + name + ('' if ok or info == '' else '  -> ' + str(info)))
    if not ok:
        fails += 1


# ── the rotation ─────────────────────────────────────────────────────────────
plan = W.order_rotation({'seeders': 70, 'newest': 15, 'random': 15, 'oldest': 0})
check('rotation is exactly 100 claims long', len(plan) == 100, len(plan))
counts = {s: plan.count(s) for s in set(plan)}
check('70/15/15 gives 70/15/15 slots', counts == {'seeders': 70, 'newest': 15, 'random': 15}, counts)
check('a zero share never appears', 'oldest' not in plan)

# The point of the interleave: any window the size of a wave of parallel fetches is already a
# proportional sample. A window shorter than the longest legitimate run of the 70 % selector can of
# course be all one kind — that is what 70 % MEANS.
worst = max(''.join('S' if p == 'seeders' else '.' for p in plan).split('.'), key=len)
check('the biggest run of one selector is short, so a wave is never one kind', len(worst) <= 6, len(worst))
for size in (8, 16, 32):
    windows = [plan[i:i + size] for i in range(0, 100 - size + 1)]
    check('every window of %d claims mixes at least two selectors' % size,
          all(len(set(w)) > 1 for w in windows))

# One percent is one claim in a hundred — small, but never "never". This is the whole reason the
# rotation is 100 long rather than, say, the parallel-fetch count.
tiny = W.order_rotation({'seeders': 99, 'random': 1})
check('a 1 % share still gets exactly one slot', tiny.count('random') == 1, tiny.count('random'))

# Deterministic: two workers reading the same settings must build the same plan, or "the order" is
# whatever the last process to start happened to compute.
check('the same shares always produce the same plan',
      W.order_rotation({'seeders': 70, 'newest': 15, 'random': 15}) ==
      W.order_rotation({'seeders': 70, 'newest': 15, 'random': 15}))

odd = W.order_rotation({'seeders': 2, 'newest': 1})
check('shares that do not total 100 still fill the rotation',
      len(odd) == 100 and 0 < odd.count('newest') < 50, odd.count('newest'))
check('every share at zero falls back to a usable plan, not an empty one',
      W.order_rotation({}) == ['oldest'] * 100)

# The whitelist takes part in the rotation like any other share.
wlplan = W.order_rotation({'whitelist': 25, 'seeders': 75})
check('the whitelist can hold a share of the rotation', wlplan.count('whitelist') == 25, wlplan.count('whitelist'))
check('…and it is spread through it, not bunched at the front',
      max(len(r) for r in ''.join('W' if p == 'whitelist' else '.' for p in wlplan).split('.')) <= 2)

# ── normalising what is in the settings table ────────────────────────────────
full = dict(W.ORDER_MIX_DEFAULT)
mode, sh = W.order_normalise('mix', {k: str(v) for k, v in full.items()})
check('a valid mix survives untouched', mode == 'mix' and sh == full, (mode, sh))
check('an unknown mode becomes the default, not a crash', W.order_normalise('sideways', {})[0] == 'oldest')
check('an empty mode becomes the default', W.order_normalise('', {})[0] == 'oldest')
# Each of these keeps ONE valid share alongside the bad one, so the "everything is zero" fallback
# does not fire and hide what is being tested.
check('nonsense in a share reads as zero, not as an exception',
      W.order_normalise('mix', {'seeders': 'lots', 'newest': '10'})[1]['seeders'] == 0)
check('a share is clamped to 0..100', W.order_normalise('mix', {'seeders': '900'})[1]['seeders'] == 100)
check('a negative share is clamped to zero',
      W.order_normalise('mix', {'seeders': '-5', 'newest': '10'})[1]['seeders'] == 0)
check('…and the valid share beside it survives',
      W.order_normalise('mix', {'seeders': '-5', 'newest': '10'})[1]['newest'] == 10)
check('mix with every share empty falls back to the default mix',
      W.order_normalise('mix', {})[1] == W.ORDER_MIX_DEFAULT)
check('normalise returns every mix key, including whitelist',
      set(W.order_normalise('mix', {})[1]) == set(W.ORDER_MIX_KEYS))

# ── the SQL each selector runs ───────────────────────────────────────────────
IDX = {'table': 'index_hashes', 'select': 'info_hash', 'gate': ' AND meta_requested_at <= NOW()',
       'key_col': 'info_hash', 'orderable': True}
WL = {'table': 'whitelist', 'select': 'id, info_hash, magnet_link', 'gate': '', 'key_col': 'id'}
q = W.Worker.claim_query.__get__(types.SimpleNamespace())

COLUMN = {'oldest': 'meta_requested_at ASC', 'newest': 'meta_requested_at DESC',
          'seeders': 'last_seeders DESC', 'seen': 'seen_count DESC', 'completed': 'last_completed DESC'}
for sel, frag in COLUMN.items():
    sql, params = q(IDX, sel)
    check('%s sorts by %s' % (sel, frag), 'ORDER BY meta_priority DESC, ' + frag in sql, sql)
    check('%s passes no parameters' % sel, params == (), params)

sql_rand, p_rand = q(IDX, 'random')
check('random seeks on the primary key instead of sorting the table',
      'info_hash >= %s' in sql_rand and 'ORDER BY info_hash' in sql_rand and 'RAND()' not in sql_rand, sql_rand)
check('…and it passes a full-width random hash',
      len(p_rand) == 1 and re.fullmatch(r'[0-9a-f]{40}', p_rand[0]) is not None, p_rand)
check('two random claims do not ask for the same point', q(IDX, 'random')[1] != q(IDX, 'random')[1])

# The expensive thing this design exists to avoid.
for sel in W.ORDER_SELECTORS:
    sql, _ = q(IDX, sel)
    check('%s never sorts by an unindexed column' % sel,
          'RAND()' not in sql and 'ORDER BY name' not in sql and 'total_size' not in sql, sql)
    check('%s still claims only ONE row' % sel, sql.rstrip().endswith('LIMIT 1'), sql)
    check('%s keeps the pending filter and the queue gate' % sel,
          "meta_status='pending'" in sql and 'meta_requested_at <= NOW()' in sql, sql)
    check('%s names an index this project actually creates' % sel,
          W.ORDER_INDEX[sel] is None or W.ORDER_INDEX[sel].startswith('idx_index_'), W.ORDER_INDEX[sel])

# The whitelist is NOT reorderable: those rows are there because a person asked for them by name.
for sel in W.ORDER_SELECTORS:
    sql, params = q(WL, sel)
    check('the whitelist ignores the "%s" selector' % sel,
          'ORDER BY meta_priority DESC, meta_requested_at ASC' in sql and params == (), sql)
    check('…and never gets an index-only column in its SQL (%s)' % sel,
          'last_seeders' not in sql and 'seen_count' not in sql and 'info_hash >=' not in sql, sql)


# ── which queue a claim goes to ──────────────────────────────────────────────
# claim_targets is the subtlest thing here: it decides whether the whitelist keeps the absolute
# priority every earlier release gave it, or takes a share. Driven directly, with no database.
def targets(mode, shares, seq=0):
    w = types.SimpleNamespace(
        queues=[WL, IDX],
        _order_seq=seq,
        effective_order=lambda: (mode, shares,
                                 W.order_rotation(shares) if mode == 'mix' else [mode] * W.ORDER_ROTATION),
    )
    return W.Worker.claim_targets(w)


for m in W.ORDER_SELECTORS:
    t = targets(m, dict(W.ORDER_MIX_DEFAULT))
    check('mode %s: the whitelist is tried FIRST' % m, t[0][0] is WL, [x[0]['table'] for x in t])
    check('mode %s: the index is tried with that ordering' % m, t[1] == (IDX, m), t[1][1])
    check('mode %s: the whitelist is claimed in queue order, never reordered' % m, t[0][1] == 'oldest')

# whitelist share 0 inside the mix = the old behaviour, unchanged
mixNoWl = {'whitelist': 0, 'seeders': 70, 'newest': 15, 'random': 15, 'oldest': 0, 'seen': 0, 'completed': 0}
t = targets('mix', mixNoWl)
check('mix with whitelist 0: the whitelist still drains first', t[0][0] is WL and t[0][1] == 'oldest',
      [(x[0]['table'], x[1]) for x in t])
check('mix with whitelist 0: no slot is ever labelled whitelist',
      'whitelist' not in W.order_rotation(mixNoWl))

# whitelist with a share: the rotation decides, and an empty queue never wastes a slot
mixWl = {'whitelist': 50, 'seeders': 50, 'newest': 0, 'random': 0, 'oldest': 0, 'seen': 0, 'completed': 0}
plan50 = W.order_rotation(mixWl)
wl_slot = plan50.index('whitelist')
idx_slot = plan50.index('seeders')
t = targets('mix', mixWl, seq=wl_slot)
check('a whitelist slot goes to the whitelist first', t[0][0] is WL, [x[0]['table'] for x in t])
check('…but still falls through to the index if the whitelist is empty', t[1][0] is IDX, len(t))
t = targets('mix', mixWl, seq=idx_slot)
check('an index slot goes to the index FIRST when the whitelist has its own share',
      t[0][0] is IDX and t[0][1] == 'seeders', [(x[0]['table'], x[1]) for x in t])
check('…and still falls through to the whitelist if the index is empty', t[1][0] is WL, len(t))
check('every claim offers both queues, so no slot is ever wasted',
      all(len(targets('mix', mixWl, seq=i)) == 2 for i in range(100)))

# The rotation must advance, or one selector would take every claim for ever.
w = types.SimpleNamespace(queues=[WL, IDX], _order_seq=0,
                          effective_order=lambda: ('mix', mixWl, plan50))
seen_slots = [W.Worker.claim_targets(w)[0][0]['table'] for _ in range(100)]
check('a hundred claims use the rotation, not slot zero a hundred times',
      len(set(seen_slots)) == 2 and 40 <= seen_slots.count('whitelist') <= 60, seen_slots.count('whitelist'))

# ── the panel and the worker must agree ──────────────────────────────────────
root = os.path.join(os.path.dirname(os.path.abspath(__file__)), '..')
schema = open(os.path.join(root, 'includes', 'schema.php'), encoding='utf-8').read()
save = open(os.path.join(root, 'api', 'admin', 'save_settings.php'), encoding='utf-8').read()
tpl = open(os.path.join(root, 'templates', 'admin', 'settings.php'), encoding='utf-8').read()
cat = open(os.path.join(root, 'includes', 'settings_catalog.php'), encoding='utf-8').read()
php = open(os.path.join(root, 'includes', 'meta_order.php'), encoding='utf-8').read()

for key in ['meta_order_mode'] + ['meta_order_mix_' + s for s in W.ORDER_MIX_KEYS]:
    check('%s has a schema default' % key, "'" + key + "'" in schema)
    check('%s is in the save allow-list' % key, "'" + key + "'" in save)
    check('%s is in the settings catalogue' % key, "'" + key + "'" in cat)
    # The field NAMES are spelled out rather than assembled at render time: the settings search finds
    # a control by its name, and the panel suite checks every catalogue key is reachable on the page.
    # A name built inside a PHP loop is invisible to both.
    if key != 'meta_order_mode':
        check('%s has a literal form field' % key, 'name="%s"' % key in tpl)


def php_list(fn):
    m = re.search(r"function %s\(\): array \{\s*return \[(.*?)\];" % fn, php, re.S)
    return tuple(re.findall(r"'(\w+)'", m.group(1))) if m else ()


def php_map(fn):
    m = re.search(r"function %s\(\): array \{\s*return \[(.*?)\];" % fn, php, re.S)
    return tuple(re.findall(r"'(\w+)'\s*=>", m.group(1))) if m else ()


check('the panel and the worker define the same selectors, in the same order',
      php_list('metaOrderSelectors') == W.ORDER_SELECTORS,
      (php_list('metaOrderSelectors'), W.ORDER_SELECTORS))
check('every mode the worker runs has a label in the form',
      set(php_map('metaOrderModeLabels')) == set(W.ORDER_MODES),
      (php_map('metaOrderModeLabels'), W.ORDER_MODES))
check('every mix key has a share field label',
      set(php_map('metaOrderShareLabels')) == set(W.ORDER_MIX_KEYS), php_map('metaOrderShareLabels'))
check('the panel and the worker distribute over the same mix keys',
      set(php_map('metaOrderDefaultMix')) == set(W.ORDER_MIX_KEYS), php_map('metaOrderDefaultMix'))

m = re.search(r"function metaOrderDefaultMix\(\): array \{\s*return \[(.*?)\];", php, re.S)
php_mix = {k: int(v) for k, v in re.findall(r"'(\w+)'\s*=> (\d+)", m.group(1))} if m else {}
check('the panel and the worker ship the SAME default mix', php_mix == W.ORDER_MIX_DEFAULT,
      (php_mix, W.ORDER_MIX_DEFAULT))
check('the default mix adds up to 100', sum(W.ORDER_MIX_DEFAULT.values()) == 100)
check('the shipped default keeps the whitelist on absolute priority',
      W.ORDER_MIX_DEFAULT['whitelist'] == 0)

m = re.search(r"function metaOrderIndexes\(\): array \{\s*return \[(.*?)\];", php, re.S)
php_idx = {}
if m:
    for k, v in re.findall(r"'(\w+)'\s*=> '?(\w+)'?", m.group(1)):
        php_idx[k] = None if v == 'null' else v
check('the panel and the worker agree on which index each ordering rides',
      php_idx == W.ORDER_INDEX, (php_idx, W.ORDER_INDEX))

# Every index named must actually be created by the schema, or the worker disables the selector for
# ever and the panel shows "building" that never finishes.
for sel, idx in W.ORDER_INDEX.items():
    if idx:
        check('schema.php creates %s (needed by "%s")' % (idx, sel), '`%s`' % idx in schema)

check('the form drives its labels and defaults from includes/meta_order.php',
      'metaOrderModeLabels()' in tpl and 'metaOrderShareLabels()' in tpl and 'metaOrderDefaultMix()' in tpl)
check('the save path normalises through the same file', 'metaOrderNormalise(' in save)
check('the form asks the database which orderings are usable', 'metaOrderAvailable(' in tpl)

print(chr(10) + '%d checks, %d failed' % (n, fails))
sys.exit(1 if fails else 0)
